src/Evolution/Systems/GeneralizedHarmonic/CharFields.cpp

```diff
@@ -0,0 +1,132 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+// IWYU pragma: no_include <array>
+
+#include "CharFields.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp" // IWYU pragma: keep
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp" // IWYU pragma: keep
+#include "Evolution/Systems/GeneralizedHarmonic/Tags.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+// IWYU pragma: no_forward_declare Tensor
+// IWYU pragma: no_forward_declare Tags::CharSpeed
+
+namespace GeneralizedHarmonic {
+
+template <size_t Dim, typename Frame>
+typename Tags::EvolvedFieldsFromCharacteristic<Dim, Frame>::type
+EvolvedFieldsFromCharacteristicCompute<Dim, Frame>::function(
+    const Scalar<DataVector>& gamma_2,
+    const tnsr::aa<DataVector, Dim, Frame>& u_psi,
+    const tnsr::iaa<DataVector, Dim, Frame>& u_zero,
+    const tnsr::aa<DataVector, Dim, Frame>& u_plus,
+    const tnsr::aa<DataVector, Dim, Frame>& u_minus,
+    const tnsr::i<DataVector, Dim, Frame>& normal_one_form) noexcept {
+  auto evolved_fields =
+      make_with_value<typename Tags::EvolvedFieldsFromCharacteristic<Dim,
+          Frame>::type>(get(gamma_2), 0.);
+
+  for (size_t mu = 0; mu < Dim + 1; ++mu) {
+    for (size_t nu = 0; nu < mu + 1; ++nu) {
+      get<::Tags::Field<Tags::Psi<Dim, Frame>>>(evolved_fields).get(mu, nu) =
+                                    u_psi.get(mu, nu);
+
+      get<::Tags::Field<Tags::Pi<Dim, Frame>>>(evolved_fields).get(mu, nu) =
+                            0.5 * (u_plus.get(mu, nu) + u_minus.get(mu, nu)) +
+                            get(gamma_2) * u_psi.get(mu, nu);
+
+      for (size_t i = 0; i < Dim; ++i) {
+        get<::Tags::Field<Tags::Phi<Dim, Frame>>>(evolved_fields).get(i,mu,nu) =
+                              0.5 * (u_plus.get(mu, nu) - u_minus.get(mu, nu)) *
+                              normal_one_form.get(i) + u_zero.get(i, mu, nu);
+      }
+    }
+  }
+
+  return evolved_fields;
+}
+
+template <size_t Dim, typename Frame>
+typename Tags::CharacteristicFields<Dim, Frame>::type
+CharacteristicFieldsCompute<Dim, Frame>::function(
+    const Scalar<DataVector>& gamma_2,
+    const tnsr::aa<DataVector, Dim>& spacetime_metric,
+    const tnsr::II<DataVector, Dim>& inverse_spatial_metric,
+    const tnsr::aa<DataVector, Dim>& pi,
+    const tnsr::iaa<DataVector, Dim>& phi,
+    const tnsr::i<DataVector, Dim, Frame>& normal_one_form) noexcept {
+  const size_t n_pts = normal_one_form.begin()->size();
+
+  // normal vector: n^i = g^{ij} n_j
+  tnsr::I<DataVector, Dim, Frame> normal{DataVector(n_pts, 0.)};
+  for (size_t i = 0; i < Dim; i++) {
+    for (size_t j = 0; j < Dim; j++) {
+       normal.get(i) += inverse_spatial_metric.get(i, j) *
+                                              normal_one_form.get(j);
+    }
+  }
```

Comment:
Instead of looping, just do `raise_or_lower_index(normal_one_form, inverse_spatial_metric)`

Reply:
done